pkg/agent/cniserver/pod_configuration.go

```diff
@@ -455,17 +250,9 @@ func (pc *podConfigurator) checkInterfaces(
 	containerID, containerNetNS, podName, podNamespace string,
```

Comment:
Question: since not many left in pod_configuration.go, do you think makes sense to move configure/remove/checkInterfaces to cniServer funcs, and just keep other non-podConfigurator funcs in this file (probably rename it)? Some funcs like validateOVSInterfaceConfig() can be changed to independent funcs.

Reply:
Also I would suggest to define an interface for OS specific funcs at least for pod_configuration funcs. It is much easier to figure out what funcs have different implementations on Windows and Linux. If we move remaining common podConfigurator funcs to cniServer, we could put other (OS specific) podConfigurator funcs to the interface implementation (but maybe rename the interface and its implementation, like interfaceConfigurator).
@tnqn: thoughts?

Reply:
AFAIK defining interface to include only OS specific funcs is not common in Go projects ([K8s](https://github.com/kubernetes/kubernetes/tree/master/pkg/kubelet/dockershim), [docker](https://github.com/moby/moby/tree/master/container)), though I think your point of easier to figure out OS specific functions make sense. But on the other side it may make the interface and its consumer logically mixed, some common part of podConfigurator is in cniServer while a few different implementation is in podConfigurator, not sure if the interface and its functions make sense in OO. 
The tradeoff is which is more important, figuring out OS specific functions via interface or the abstract and encapsulation of an interface, personally I think the _OS.go files themselves are clear to figure out OS specific functions too, so prefer keeping an interface logically complete and leverage Go build constraints to take care of the OS difference.

Reply:
I understand what you mean. However for cases like interface configuration, route configuration, iptables (traffic filter), seems they can be naturally mapped to an interface/class, and they do not share code (except some utility code does not really need a receiver) in Windows and Linux, as they are really OS interfaces.

If we do not want to create an interface, we can also just do this - let us revisit the remaining funcs in pod_configurator.go, to see if they should be kept in PodConfigurator, or it is good to move them to cniServer or remove the receiver.
To me it looks strange that a class is split to xyz.go and xyz_os.go, while xyz.go does not really include much common code (if xyz.go really has many common funcs of the class, while xyz_os.go just includes a small portion that is different in different OSes, it sounds better to me).

Reply:
I'm not against creating an interface, just prefer the interface is object centralized, not customized for OS specific functions, like podConfigurator with configureInterfaces, removeInterfaces and checkInterfaces functions is a complete abstract to me.
I'm not sure what abstract and functions we will have by extracting some OS specific code, but I'm fine with it as long as it can be abstracted naturally.
The only question is if the new object has to share some code between different OS implementation, is there a better way than putting the common code to xyz.go?
From my understanding, abc.go and abc_linux.go should be considered a whole as their names implies, the suffix is just a facility Go provides to leverage the build constraints.

Reply:
For the shared code, I am thinking if they are actually some utility code - that means they are not quite relevant to the PodConfigurator object, but can be standalone funcs (no receiver), like findContainerIPConfig, validateOVSInterfaceConfig, parseContainerIP, etc.; and there are other funcs like configureInterfaces, removeInterfaces, checkInterfaces which we might just move to cniServer.

Reply:
@jianjuns   In my opinion, currently cniserver is taking these tasks, 1) communicating with CNI plugin, 2) calling IPAM to allocate/release IP, 3) calling podConfigurator to configure/check container networking interface. If we move functions in podConfigurator to cniserver, it would break the encapsulations. Besides, if we split podConfigurators, some fields need to be existed in two structs(cniserver and podConfigurator), e.g., ovsBridgeClient, ofClient, ifaceStore. etc. I think this should not be as expected.

Reply:
@wenyingd : I think your point is valid that cniServer now does not interact with OVS, etc..

I looked into the pod_configuration_os code further, and found actually it does not interact with ovsdb (except OVS port creation which we might be able to move the shared code. See my comment on createOVSPort()), openflow, and interfaceStore either. So: 1) it is really code specific to link manipulation, separate from the rest of the shared podConfigurator; 2) it does not really share state with podConfigurator. I kind of believe it makes sense to move the code out of podConfigurator, like interface_configuraiton_os.go. We can decide whether or not to create a interface separately (if no interface, for Linux we might not even need a struct as seems no saved state needs to passed across funcs, but I understand for Windows we might have some shared state to keep in a struct).

What you think.

Reply:
I think I get your point. I will rename current pod_configuration_xx.go to interface_configuration_os.go. And I could try to define a new interface "interfaceConfigurator", which could be called in `podConfigurator`, and implement this new interface in interface_configuration_os.go. Is it correct?